kafka-rest/src/main/java/io/confluent/kafkarest/response/StreamingResponse.java

```diff
@@ -141,31 +154,56 @@ private ResultOrError handleNext(T result, @Nullable Throwable error) {
   private static class InputStreamingResponse<T> extends StreamingResponse<T> {
 
     private final MappingIterator<T> mappingIteratorInput;
+    private final StreamingResponseIdleTimeLimiter idleTimeLimiter;
+
+    private Throwable hasNextError = null;
+    private boolean closed = false;
 
     private InputStreamingResponse(
-        MappingIterator<T> mappingIteratorInput, ChunkedOutputFactory chunkedOutputFactory) {
+        MappingIterator<T> mappingIteratorInput,
+        ChunkedOutputFactory chunkedOutputFactory,
+        StreamingResponseIdleTimeLimiter idleTimeLimiter) {
       super(chunkedOutputFactory);
       this.mappingIteratorInput = requireNonNull(mappingIteratorInput);
+      this.idleTimeLimiter = requireNonNull(idleTimeLimiter);
     }
 
     public void close() {
-      try {
-        mappingIteratorInput.close();
-      } catch (IOException e) {
-        log.error("Error when closing the request stream.", e);
-      }
+      closed = true;
     }
 
     @Override
     public boolean hasNext() {
-      return mappingIteratorInput.hasNext();
+      if (closed) {
+        return false;
+      }
+      try {
+        return idleTimeLimiter.callWithTimeout(mappingIteratorInput::hasNextValue);
+      } catch (ExecutionException | UncheckedExecutionException e) {
+        hasNextError = e.getCause();
+        return true;
+      } catch (InterruptedException | MaxIdleTimeExceededException e) {
+        hasNextError = e;
+        return true;
+      }
     }
 
     @Override
     public CompletableFuture<T> next() {
+      if (closed) {
+        throw new NoSuchElementException();
+      }
+      if (hasNextError != null) {
+        close();
+        return CompletableFutures.failedFuture(hasNextError);
+      }
       try {
-        return CompletableFuture.completedFuture(mappingIteratorInput.nextValue());
-      } catch (Throwable e) {
+        return CompletableFuture.completedFuture(
+            idleTimeLimiter.callWithTimeout(mappingIteratorInput::nextValue));
+      } catch (ExecutionException | UncheckedExecutionException e) {
+        return CompletableFutures.failedFuture(e.getCause());
```

Comment:
Shouldn't we also close the stream here? What are the cases that should return an error without closing the stream? I was thinking about handling bad JSON, where arguably it might be OK to keep the stream open, but wouldn't that practically always be caught in `hasNext()`, where execution exceptions do close the stream (by setting `hasNextError`)? This is all pretty subtle, so can we please add generous comments around it?